File: crates/worker-sdk/src/handler.rs
//! TaskHandler trait and TaskContext (design §1). User implements TaskHandler; runtime passes TaskContext.

use async_trait::async_trait;
use std::time::Duration;

use crate::client::EngineClient;
use crate::types::{ExternalTask, TaskResult};

/// Context passed to the handler. Provides access to task metadata and
/// the ability to extend the lock for long-running tasks.
#[derive(Clone)]
pub struct TaskContext {
    pub(crate) worker_id: String,
    pub(crate) task_id: String,
    pub(crate) client: EngineClient,
}

impl TaskContext {
    pub fn new(worker_id: String, task_id: String, client: EngineClient) -> Self {
        Self {
            worker_id,
            task_id,
            client,
        }
    }

    /// The worker ID that owns this task's lock.
    pub fn worker_id(&self) -> &str {
        &self.worker_id
    }

    /// The task ID.
    pub fn task_id(&self) -> &str {
        &self.task_id
    }

    /// Extend the lock on this task to prevent timeout during long processing.
    ///
    /// Call this periodically for tasks that may exceed the initial lock duration.
    pub async fn extend_lock(&self, extension: Duration) -> Result<(), crate::client::ClientError> {
        self.client
            .extend_lock(&self.task_id, &self.worker_id, extension)
            .await
    }
}

/// User implements this to handle tasks (design §1).
#[async_trait]
pub trait TaskHandler: Send + Sync {
    fn task_type(&self) -> &str;

    async fn handle(&self, task: ExternalTask, ctx: TaskContext) -> TaskResult;
}

File: sdk/python/bpm_engine_sdk/handler.py
"""TaskHandler abstract base class and TaskContext."""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import TYPE_CHECKING

from bpm_engine_sdk.models import ExternalTask, TaskResult

if TYPE_CHECKING:
    from bpm_engine_sdk.client import EngineClient


@dataclass
class TaskContext:
    """Context passed to a task handler during execution.

    Provides access to task metadata and the ability to extend the lock
    for long-running tasks.
    """

    worker_id: str
    task_id: str
    _client: EngineClient | None = None

    async def extend_lock(self, extension_ms: int) -> None:
        """Extend the lock on this task to prevent timeout.

        Call this periodically for tasks that may exceed the initial
        lock duration.

        Args:
            extension_ms: Extension duration in milliseconds.

        Raises:
            RuntimeError: If no client was configured.
            EngineError: If the engine rejects the request.
        """
        if self._client is None:
            raise RuntimeError("extend_lock requires an EngineClient")
        await self._client.extend_lock(self.task_id, self.worker_id, extension_ms)


class TaskHandler(ABC):
    """Abstract base class for external task handlers.

    Implement ``task_type`` and ``handle`` to process tasks of a specific type.

    Example::

        class PaymentHandler(TaskHandler):
            @property
            def task_type(self) -> str:
                return "payment"

            async def handle(self, task: ExternalTask, ctx: TaskContext) -> TaskResult:
                amount = task.variables.get("amount", "0")
                # process payment ...
                return TaskResult.complete({"status": "paid"})
    """

    @property
    @abstractmethod
    def task_type(self) -> str:
        """The task type this handler processes (matches the BPMN topic)."""
        ...

    @abstractmethod
    async def handle(self, task: ExternalTask, ctx: TaskContext) -> TaskResult:
        """Handle an external task and return the result."""
        ...
